### PR TITLE
NIFI-4644 Fixed LookupService API to allow for more than String/Strin…

### PR DESCRIPTION
**Pull request by MikeThomsen**
…g lookup pairs.

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by MikeThomsen** [comment]
@m-hogue I think on the mailing list it said you did a lot of work on the Record API. Do you think you could take a look at this? It's an API change that was necessitated by the String,String limitation.

File: nifi-nar-bundles/nifi-standard-services/nifi-hbase_1_1_2-client-service-bundle/nifi-hbase_1_1_2-client-service/src/main/java/org/apache/nifi/hbase/HBase_1_1_2_RecordLookupService.java [hunk 1]
```diff
@@ -114,8 +114,12 @@
...
 
     @Override
-    public Optional<Record> lookup(Map<String, String> coordinates) throws LookupFailureException {
-        final String rowKey = coordinates.get(ROW_KEY_KEY);
+    public Optional<Record> lookup(Map<String, Object> coordinates) throws LookupFailureException {
+        if (coordinates.get(ROW_KEY_KEY) == null) {
+            return Optional.empty();
+        }
+
+        final String rowKey = coordinates.get(ROW_KEY_KEY) instanceof String ? (String)coordinates.get(ROW_KEY_KEY) : coordinates.get(ROW_KEY_KEY).toString();
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-services/nifi-hbase_1_1_2-client-service-bundle/nifi-hbase_1_1_2-client-service/src/main/java/org/apache/nifi/hbase/HBase_1_1_2_RecordLookupService.java]: I don't think we need to be checking any instanceof here -- just call coordinates.get(ROW_KEY_KEY).toString() -- it is a String, then toString() will simply return 'this'.

File: nifi-nar-bundles/nifi-standard-services/nifi-lookup-services-bundle/nifi-lookup-services/src/main/java/org/apache/nifi/lookup/SimpleCsvFileLookupService.java [hunk 2]
```diff
@@ -194,12 +194,12 @@ public void onEnabled(final ConfigurationContext context) throws InitializationE
...
 
     @Override
-    public Optional<String> lookup(final Map<String, String> coordinates) throws LookupFailureException {
+    public Optional<String> lookup(final Map<String, Object> coordinates) throws LookupFailureException {
         if (coordinates == null) {
             return Optional.empty();
         }
 
-        final String key = coordinates.get(KEY);
+        final String key = coordinates.get(KEY) instanceof String ? (String)coordinates.get(KEY) : coordinates.get(KEY).toString();
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-services/nifi-lookup-services-bundle/nifi-lookup-services/src/main/java/org/apache/nifi/lookup/SimpleCsvFileLookupService.java]: Again, no need to check instanceof here - can just call toString()

File: nifi-nar-bundles/nifi-standard-services/nifi-lookup-services-bundle/nifi-lookup-services/src/main/java/org/apache/nifi/lookup/SimpleKeyValueLookupService.java [hunk 3]
```diff
@@ -58,12 +58,12 @@ public void cacheConfiguredValues(final ConfigurationContext context) {
...
 
     @Override
-    public Optional<String> lookup(final Map<String, String> coordinates) {
+    public Optional<String> lookup(final Map<String, Object> coordinates) {
         if (coordinates == null) {
             return Optional.empty();
         }
 
-        final String key = coordinates.get(KEY);
+        final String key = coordinates.get(KEY) instanceof String ? (String)coordinates.get(KEY) : coordinates.get(KEY).toString();
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-services/nifi-lookup-services-bundle/nifi-lookup-services/src/main/java/org/apache/nifi/lookup/SimpleKeyValueLookupService.java]: Again, no need to check instanceof here - can just call toString()

File: nifi-nar-bundles/nifi-standard-services/nifi-lookup-services-bundle/nifi-lookup-services/src/main/java/org/apache/nifi/lookup/configuration2/CommonsConfigurationLookupService.java [hunk 4]
```diff
@@ -122,12 +122,12 @@ public void onEvent(ConfigurationBuilderEvent event) {
...
 
     @Override
-    public Optional<String> lookup(final Map<String, String> coordinates) throws LookupFailureException {
+    public Optional<String> lookup(final Map<String, Object> coordinates) throws LookupFailureException {
         if (coordinates == null) {
             return Optional.empty();
         }
 
-        final String key = coordinates.get(KEY);
+        final String key = coordinates.get(KEY) instanceof String ? (String)coordinates.get(KEY) : coordinates.get(KEY).toString();
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-services/nifi-lookup-services-bundle/nifi-lookup-services/src/main/java/org/apache/nifi/lookup/configuration2/CommonsConfigurationLookupService.java]: Again, no need to check instanceof here - can just call toString()

File: nifi-nar-bundles/nifi-standard-services/nifi-lookup-services-bundle/nifi-lookup-services/src/main/java/org/apache/nifi/lookup/maxmind/IPLookupService.java [hunk 5]
```diff
@@ -237,8 +237,8 @@ public void closeReader() throws IOException {
         }
     }
 
-    private Optional<Record> doLookup(final DatabaseReader databaseReader, final Map<String, String> coordinates) throws LookupFailureException, InvalidDatabaseException {
-        final String ipAddress = coordinates.get(IP_KEY);
+    private Optional<Record> doLookup(final DatabaseReader databaseReader, final Map<String, Object> coordinates) throws LookupFailureException, InvalidDatabaseException {
+        final String ipAddress = coordinates.get(IP_KEY) instanceof String ? (String)coordinates.get(IP_KEY) : coordinates.get(IP_KEY).toString();
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-services/nifi-lookup-services-bundle/nifi-lookup-services/src/main/java/org/apache/nifi/lookup/maxmind/IPLookupService.java]: Again, no need to check instanceof here - can just call toString()

**Comment by MikeThomsen** [comment]
Thanks @markap14 I'll fix these and push a new version.

**Comment by markap14** [comment]
@MikeThomsen I am just a bit concerned about potential implications of this change... at first glance I don't see any particular problem with it, but I feel like there may be a condition that I am overlooking... It is probably okay, but I would like to think about this just a bit more.

**Comment by MikeThomsen** [comment]
@markap14 That's fine. I posted on the mailing list that someone already found a real case where it broke their flow because it converted an integer to a string and MongoDB couldn't do the lookup. The ticket reference for that particular problem is in there if you want to take a look.

**Comment by MikeThomsen** [comment]
Ok, code review notes should all be addressed now.

**Comment by MikeThomsen** [comment]
@markap14 Have you had a chance to think more this? FWIW, it doesn't seem to break any of the unit tests.

**Pull request closed without merging**